Hw 38 solution

ipantazi opened the pull request.

ipantazi committed:
Added `Book` entity, class `DataProcessingException`, `BookRepository` interface with methods and `BookRepositoryImpl` class.
Added `BookService` interface with methods and `BookServiceImpl` class.
Modified class `Application` and file `application.properties`.
Created a new folder `resources` in the `src/test` directory
Create a new file `application.properties` in the `src/test/resources` folder.
Added dependencies to the `pom.xml` file.

ipantazi committed:
Using EntityManagerFactory instead of SessionFactory in `BookRepositoryImpl` class

ipantazi committed:
Deleted dependency `commons-dbcp` in `pom.xml`.
Set upped all fields and renamed object in class `Application`.
 Renamed schema in `application.properties`.
 Added annotation `+@RequiredArgsConstructor` and deleted constructor in classes `BookServiceImpl` and `BookRepositoryImpl`.
 Added required constraints for fields, changed annotation `@Data` to `@Getter`, `@Setter`, `@Builder`, `@ToString`, `@NoArgsConstructor`, `@AllArgsConstructor`, added annotation `@Table` and methods `equals` and `hashCode` in class `Book`.

ipantazi committed:
Removed methods: `equals()`, hashCode();
annotations: `@ToString`, `@NoArgsConstructor`, `@AllArgsConstructor`, `@Builder`;
 modified annotations `@Column` for all fields in `Book` class.

ipantazi committed:
Removed CommandLineRunner bean from `Application` class.
Created classes: `BookController`, `BookDto`, `CreateBookRequestDto`, `EntityNotFoundException`, `MapperConfig` and interface `BookMapper`.
Added method `getBookById(Long id)` to classes: `BookRepositoryImpl`, `BookServiceImpl` and interface `BookRepository`.
Added dependencies to `pom.xml`.

ipantazi committed:
Added to `application.properties` class - `spring.jpa.open-in-view=false` because was `WARN 7132 spring.jpa.open-in-view is enabled by default`.
Added to `BookMapper` interface - `@Mapping(target = "id", ignore = true)` because was `[WARNING] Unmapped target property: "id"`.

ipantazi committed:
Added version lombok to the dependency in the `pom.xml`.

ipantazi committed:
Modified dependency lombok in the `pom.xml`.

ipantazi committed:
Removed the annotation `@Column(nullable = false)` from the class `Book`.
The interface `BookRepository` is extended by the interface `JpaRepository`, all methods from this interface are removed and the class `BookRepositoryImpl` is removed.

ipantazi committed:
Added annotation `@ResponseStatus(HttpStatus.CREATED)` to method `createBook()` of class `BookController`.
Added files: `01-create-books-table.yaml` and `db.changelog-master.yaml`. Changed line `spring.jpa.hibernate.ddl-auto=validate` in file `application.properties`.
Added field `deleted` to class `Book`.
Added method `updateBookById` to class `BookRepository`.
Added methods: `deleteById` and `update` to class `BookRepository`.
Added methods: `delete` and `updateBook` to class `BookRepository`.
Added dependency `liquibase` to file `pom.xml`.

liliia-ponomarenko commented on src/main/java/mate/academy/onlinebookstore/controller/BookController.java:
```diff
@@ -0,0 +1,53 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping(value = "/books")
```
```suggestion
@RequestMapping("/books")
```

liliia-ponomarenko commented on src/main/java/mate/academy/onlinebookstore/controller/BookController.java:
```diff
@@ -0,0 +1,53 @@
+
+    @PutMapping("/{id}")
+    @ResponseStatus(HttpStatus.OK)
+    public int updateBook(@PathVariable Long id,
```
```suggestion
    public BookDto updateBook(@PathVariable Long id,
```
Better return updated book

liliia-ponomarenko commented on src/main/java/mate/academy/onlinebookstore/model/Book.java:
```diff
@@ -0,0 +1,46 @@
+    private String coverImage;
+
+    @Column(name = "is_deleted", nullable = false)
+    private boolean deleted = false;
```
```suggestion
    private boolean isDeleted = false;
```

liliia-ponomarenko commented on src/main/java/mate/academy/onlinebookstore/repository/BookRepository.java:
```diff
@@ -0,0 +1,25 @@
+
+public interface BookRepository extends JpaRepository<Book, Long> {
+
+    @Modifying
+    @Transactional
+    @Query(value = "UPDATE Book SET title = :title, author = :author, isbn = :isbn, "
+            + "price = :price, description = :description, coverImage = :coverImage WHERE id = :id "
+            + "AND deleted = false")
+    int updateBookById(@Param("id") Long id,
+                        @Param("title") String title,
+                        @Param("author") String author,
+                        @Param("isbn") String isbn,
+                        @Param("price") BigDecimal price,
+                        @Param("description") String description,
+                        @Param("coverImage") String coverImage);
```
Why do you need it? Can you use the method save CrudRepository?

liliia-ponomarenko commented on src/main/java/mate/academy/onlinebookstore/service/BookServiceImpl.java:
```diff
@@ -0,0 +1,54 @@
+
+    @Override
+    public int update(Long id, CreateBookRequestDto bookRequestDto) {
+        return bookRepository.updateBookById(id,
+                bookRequestDto.getTitle(),
+                bookRequestDto.getAuthor(),
+                bookRequestDto.getIsbn(),
+                bookRequestDto.getPrice(),
+                bookRequestDto.getDescription(),
+                bookRequestDto.getCoverImage());
```
Use Mapper for it;) Check if the book exists before updating

liliia-ponomarenko commented on src/main/java/mate/academy/onlinebookstore/service/BookServiceImpl.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    @Override
+    public void deleteById(Long id) {
```
the same. Let’s check that the book exists by id.

liliia-ponomarenko commented on src/main/resources/db/changelog/changes/01-create-books-table.yaml:
```diff
@@ -0,0 +1,53 @@
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
+
```
leave one empty line at the end of the file

liliia-ponomarenko reviewed (changes requested):
Good job! Let’s fix a few mistakes ;)

ipantazi committed:
Deleted the cod "value" and the returned object has been changed in the method `updateBook()` in the class `BookController`.
Renamed field "deleted" in the class `Book`.
Deleted the method `updateBookById()` from interface `BookRepository`.
Leave one empty line at the end of the file `01-create-books-table.yaml`.
Added the new method `updateBookFromDto()` to the interface `BookMapper`.
Changed the methods: `findById()`, `update()`, `deleteById()` and added new private method `findBookById()`.

Elena-Bruyako commented on src/main/java/mate/academy/onlinebookstore/model/Book.java:
```diff
@@ -0,0 +1,46 @@
+@Entity
+@Table(name = "books")
+@SQLDelete(sql = "UPDATE books SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
```
check if `@Where` is deprecated or not

Elena-Bruyako commented on src/main/java/mate/academy/onlinebookstore/model/Book.java:
```diff
@@ -0,0 +1,46 @@
+    @Column(nullable = false)
+    private String description;
+
+    @Column(name = "cover_image", nullable = false)
+    private String coverImage;
+
+    @Column(name = "is_deleted", nullable = false)
```
```suggestion
    @Column(nullable = false)
    private String coverImage;

    @Column(nullable = false)
```

Elena-Bruyako commented on src/main/java/mate/academy/onlinebookstore/service/BookServiceImpl.java:
```diff
@@ -0,0 +1,53 @@
+
+    @Override
+    public BookDto update(Long id, CreateBookRequestDto bookRequestDto) {
+        Book updatedBook = bookMapper.updateBookFromDto(bookRequestDto, findBookById(id));
+        bookRepository.save(updatedBook);
```
```suggestion
        Book book = findBookById(Long id);
        bookMapper.updateBookFromDto(bookRequestDto, book);
        return bookMapper.toDto(bookRepository.save(book));
```

Elena-Bruyako commented on src/main/java/mate/academy/onlinebookstore/mapper/BookMapper.java:
```diff
@@ -0,0 +1,22 @@
+
+    @Mapping(target = "id", ignore = true)
+    @Mapping(target = "deleted", ignore = true)
+    Book updateBookFromDto(CreateBookRequestDto bookRequestDto, @MappingTarget Book book);
```
```suggestion
    void updateBookFromDto(CreateBookRequestDto bookRequestDto, @MappingTarget Book book);
```

Elena-Bruyako reviewed (changes requested):
see comments

ipantazi committed:
Changed the implementation of "soft deleted" in the `Book` class.
Fixed the code in the `update()` method in the `BookServiceImpl` class.
Changed the return value in the `updateBookFromDto` method in the `BookMapper` interface to "void".

Elena-Bruyako commented on src/main/java/mate/academy/onlinebookstore/model/Book.java:
```diff
@@ -0,0 +1,44 @@
+    @Column(nullable = false)
+    private String description;
+
+    @Column(nullable = false)
```
```suggestion
```

Elena-Bruyako commented on src/main/java/mate/academy/onlinebookstore/model/Book.java:
```diff
@@ -0,0 +1,44 @@
+    @Column(nullable = false)
+    private BigDecimal price;
+
+    @Column(nullable = false)
```
```suggestion
```

Elena-Bruyako reviewed (changes requested):
two comments

ipantazi committed:
Removed `@Column(nullable = false)` annotation from `description`, `coverImage` and `isDeleted` fields in `Book` class.

Elena-Bruyako commented on src/main/resources/db/changelog/changes/01-create-books-table.yaml:
```diff
@@ -0,0 +1,52 @@
+              - column:
+                  name: cover_image
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
```
```suggestion
```

Elena-Bruyako commented on src/main/resources/db/changelog/changes/01-create-books-table.yaml:
```diff
@@ -0,0 +1,52 @@
+              - column:
+                  name: description
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
```
```suggestion
```

Elena-Bruyako reviewed (changes requested):
don't forget about changesets when you make some changes in entity

ipantazi committed:
Removed constraints from columns `cover_image` and `description`.
Changed the implementation of "soft deleted" in the `Book` class.

liliia-ponomarenko commented on src/main/java/mate/academy/onlinebookstore/dto/BookDto.java:
```diff
@@ -0,0 +1,19 @@
+
+@Getter
+@Setter
+@RequiredArgsConstructor
```
Why do you need this annotation here?

liliia-ponomarenko commented on src/main/java/mate/academy/onlinebookstore/dto/CreateBookRequestDto.java:
```diff
@@ -0,0 +1,18 @@
+
+@Getter
+@Setter
+@RequiredArgsConstructor
```
the same

liliia-ponomarenko reviewed (approved):
Good job! A few minor comments;)
